MPR#6567: add Unix.flush_mapped_file

xavierleroy opened the pull request:
This function synchronizes a memory-mapped file with the corresponding in-memory bigarray.

This PR reuses the code proposed by Christoph Bauer in [MPR#6567](https://caml.inria.fr/mantis/view.php?id=6567) and adapts it to the new division of labor between Bigarray and Unix, whereas functions for memory-mapped files are now in Unix.

xavierleroy committed:
MPR#6567: add Unix.flush_mapped_file

This function synchronizes a memory-mapped file with the corresponding in-memory bigarray.

avsm commented on otherlibs/unix/mmap_ba.c:
```diff
@@ -89,3 +89,26 @@ ALLOC_FUNCTION(int flags, int num_dims, void * data, intnat * dim)
+  if( b->flags & CAML_BA_MAPPED_FILE ) {
+    if( b->proxy == NULL )
+      caml_unix_flush_file(b->data, caml_ba_byte_size(b), sync);
+    else {
```
A minor style nit, but one of the single-line branches here has a pair of braces, and the other does not.

xavierleroy replied:
Minor style changes applied for consistency with the rest of the C code base.

avsm commented on otherlibs/unix/unix.mli:
```diff
@@ -543,6 +543,28 @@ val map_file :
   validation fails.
   @since 4.06.0 *)
 
+val flush_mapped_file :
```
Does `UnixLabels` also need a corresponding function added to it?

xavierleroy replied:
I'm afraid so.  Added `UnixLabels.flush_mapped_file`.

chrbauer commented on otherlibs/unix/mmap.c:
```diff
@@ -225,3 +225,33 @@ void UNMAP_FILE_FUNCTION(void * addr, uintnat len)
+CAMLexport void caml_unix_flush_file(void * addr, uintnat len, int sync)
+{
+  caml_invalid_argument("Unix.flush_mapped_file: not supported");
+  return Val_unit;
```
A void-function should not return Val_unit. I would expect a compile error

xavierleroy replied:
Well spotted.  Those fallback implementations are rarely exercised.

xavierleroy committed:
MPR#6567: add Unix.flush_mapped_file, continued

- Type error in fallback implementation of caml_unix_flush_file
- Added flush_mapped_file to the UnixLabels interface
- C style tweaks

xavierleroy commented:
There is a new commit on my branch that doesn't show up here.  https://github.com/xavierleroy/ocaml/commit/ed03e8b35ce82d08f2e5ed45d268d5173660757a .  Hope it will show up eventually.

dra27 commented on otherlibs/win32unix/mmap.c:
```diff
@@ -161,6 +161,32 @@ void UNMAP_FILE_FUNCTION(void * addr, uintnat len)
+  int ret;
+
+  if (sync)
+    caml_invalid_argument("Unix.flush_mapped_file: sync is not supported under windows");
```
Micro-nit: capital 'W' on Windows

dra27 reviewed (commented):
I've never played with the memory mapped stuff, but I was just looking into the feasibility of using FlushFileBuffers for the `sync` case (it would potentially do more work than `MS_SYNC`, but would basically match the semantics).

However, that would require the original handle to be available (a bit of work, but I think that could be safely done when the view is created). What surprised me is that the file mapping handle `fmap` is immediately closed after the view of the created - I don't think that's sound?

xavierleroy commented:
> What surprised me is that the file mapping handle fmap is immediately closed after the view of the created - I don't think that's sound?

Quoting the MSDN page for CreateFileMapping:

Mapped views of a file mapping object maintain internal references to the object, and a file mapping object does not close until all references to it are released. Therefore, to fully close a file mapping object, an application must unmap all mapped views of the file mapping object by calling UnmapViewOfFile and close the file mapping object handle by calling CloseHandle. **These functions can be called in any order.**

(Emphasis mine.)

dra27 commented:
Ah, my bad, sorry! What do you think of adding a mapping from base addresses back to file handles so that the sync case can be implemented?

xavierleroy commented:
> What do you think of adding a mapping from base addresses back to file handles so that the sync case can be implemented?

It sounds painful to maintain that mapping...  However, we could possibly ask the programmer to do the work for us by taking the original file descriptor as extra argument of `flush_mapped_file`:
```
val flush_mapped_file:
  file_descr -> ('a, 'b, 'c) CamlinternalBigarray.genarray -> bool -> unit
```
On the one hand it's not an excessive burden on the programmer.  On the other hand it raises nasty questions of the kind "what if the file descriptor and the bigarray don't match?", with answers such as "maybe no flush will happen, or some other file will be flushed".

alainfrisch commented on otherlibs/unix/unix.mli:
```diff
@@ -543,6 +543,28 @@ val map_file :
+   the underlying file descriptor initially passed to {!Unix.map_file}
+   is closed; otherwise, depending on the OS and file system, there is
+   a risk that the changes made to [b] are lost.
+
```
> and not update the file for arbitrary long amounts of time

When `sync = false`, I assume there is no guarantee about how long it could take for the synchronization to complete, right?  Is is still guarantee that close the file descriptor immediately after calling `flush_mapped_filed` is safe?

alainfrisch commented on otherlibs/unix/unix.mli:
```diff
@@ -543,6 +543,28 @@ val map_file :
   @since 4.06.0 *)
 
+val flush_mapped_file :
+  ('a, 'b, 'c) CamlinternalBigarray.genarray -> bool -> unit
```
I'm in favor of labeling the bool argument (even in Unix); this makes the client code more readable (the meaning of the bool is really not obvious without looking at the documentation).

alainfrisch commented on otherlibs/unix/mmap_ba.c:
```diff
@@ -89,3 +89,26 @@ ALLOC_FUNCTION(int flags, int num_dims, void * data, intnat * dim)
+  struct caml_ba_array * b = Caml_ba_array_val(vb);
+  int sync = Bool_val(vsync);
+  
+  if (b->flags & CAML_BA_MAPPED_FILE) {
```
Shouldn't the function raise an exception when the bigarray is not on a mapped file?  Or do you envision cases where some code would need to flush a bigarray without knowing itself if it is mapped?

alainfrisch commented on otherlibs/win32unix/mmap.c:
```diff
@@ -161,6 +161,32 @@ void UNMAP_FILE_FUNCTION(void * addr, uintnat len)
+  caml_enter_blocking_section();
+  ret = FlushViewOfFile(addr, 0);
+       /* zero means: the file is flushed from the base 
+          address to the end of the mapping. */
```
The Unix version takes care of passing the length of the area to flush.  Is there a reason not to do the same here?

alainfrisch commented:
> What do you think of adding a mapping from base addresses back to file handles so that the sync case can be implemented?

Cannot we replace the CAML_BA_MAPPED_FILE by an extra field in caml_ba_array to hold the handle?  Are you concerned by the cost of adding such as field to all big array descriptors?  If so, it looks possible to turn the proxy field into a union (of the current proxy and the handle), controlled by CAML_BA_MAPPED_FILE (we would follow proxies to find the handle, so no need to keep both a proxy structure and the handle).

xavierleroy commented:
Two years later, I completely forgot about this PR and have swapped out all context.  Reading the discussion again, I can't understand whether the reviewers were happy with the final version or not, and what changes they wanted.  (As opposed to "wouldn't it be cool to do it differently?" comments.)  So, I'm closing this PR.  Whoever cares enough about the issue is welcome to pick up my code and work from that.
